### PR TITLE
Job, task, taskflow log notifications

### PR DESCRIPTION
**Pull request by cjh1**

### CODE REVIEW

File: girder/cumulus/plugin_tests/cluster_test.py [hunk 1]
```diff
@@ -973,33 +973,3 @@ def test_create_scheduler_type(self, generate_key_pair):
...
-        log_entry = {
-            'msg': 'Some message'
-        }
-        r = self.request('/clusters/%s/log' % str(cluster_id), method='POST',
-                         type='application/json', body=json.dumps(log_entry), user=self._user)
-        self.assertStatusOk(r)
-
-        notifications = self.getSseMessages(stream_r)
-        self.assertEqual(len(notifications), 1, 'Expecting one notification')
-        self.assertEqual(notifications[0]['msg'], log_entry['msg'], 'Expecting a different log message')
```
Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: Where has this test code gone?

Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: I guess this endpoint has been removed? Makes sense then!

Comment (cjh1) [inline girder/cumulus/plugin_tests/cluster_test.py]: Would be nice to test that we get the events on the /notifications endpoint ?

Comment (TristanWright) [inline girder/cumulus/plugin_tests/cluster_test.py]: will add some event stream tests ✓

**Comment by cjh1** [comment]
I just tested this branch and run into a couple of issues.

- The cluster log no longer updates automatically every time ( a page refresh is sometimes required ).
- There seem to be duplicate log entries displayed see screenshot below:
duplicate

**Comment by TristanWright** [comment]
For the duplicate logs, was it on the cluster, task, job, or taskflow?

**Comment by cjh1** [comment]
The duplicate entries where on the cluster.

**Comment by TristanWright** [comment]
Looks like we're appending twice somewhere. For some though, the entries are basically identical except the one has stdout. 

```json
  {
    "args": [
      
    ],
    "created": 1467069778.5395,
    "data": {
      "_ansible_no_log": false,
      "cache_update_time": 0,
      "cache_updated": false,
      "changed": false,
      "module_name": "apt"
    },
    "exc_info": null,
    "exc_text": null,
    "filename": "cumulus_log.py",
    "funcName": "log",
    "levelname": "INFO",
    "levelno": 20,
    "lineno": 60,
    "module": "cumulus_log",
    "msecs": 539.50500488281,
    "msg": "TASK: exec : Install NFS client",
    "name": "cumulus_log",
    "pathname": "\/opt\/hpccloud\/cumulus\/cumulus\/ansible\/tasks\/playbooks\/callback_plugins\/cumulus_log.py",
    "process": 31766,
    "processName": "MainProcess",
    "relativeCreated": 357778.70893478,
    "status": "finished",
    "thread": 1.3965367359674e+14,
    "threadName": "MainThread",
    "type": "task"
  },
  {
    "args": [
      
    ],
    "created": 1467069793.284,
    "data": {
      "_ansible_no_log": false,
      "cache_update_time": 0,
      "cache_updated": false,
      "changed": true,
      "module_name": "apt",
      "stderr": "\nCreating config file \/etc\/idmapd.conf with new version\n\nCreating config file \/etc\/default\/nfs-common with new version\n",
      "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following extra packages will be installed:\n  keyutils libgssglue1 libnfsidmap2 libtirpc1 rpcbind\nSuggested packages:\n  open-iscsi watchdog\nThe following NEW packages will be installed:\n  keyutils libgssglue1 libnfsidmap2 libtirpc1 nfs-common rpcbind\n0 upgraded, 6 newly installed, 0 to remove and 29 not upgraded.\nNeed to get 375 kB of archives.\nAfter this operation, 1524 kB of additional disk space will be used.\nGet:1 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libgssglue1 amd64 0.4-2ubuntu1 [19.7 kB]\nGet:2 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libnfsidmap2 amd64 0.25-5 [32.2 kB]\nGet:3 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libtirpc1 amd64 0.2.2-5ubuntu2 [71.3 kB]\nGet:4 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main keyutils amd64 1.5.6-1 [33.6 kB]\nGet:5 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty-updates\/main rpcbind amd64 0.2.1-2ubuntu2.2 [37.1 kB]\nGet:6 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty-updates\/main nfs-common amd64 1:1.2.8-6ubuntu1.2 [181 kB]\nFetched 375 kB in 0s (6884 kB\/s)\nSelecting previously unselected package libgssglue1:amd64.\n(Reading database ... 74563 files and directories currently installed.)\nPreparing to unpack ...\/libgssglue1_0.4-2ubuntu1_amd64.deb ...\nUnpacking libgssglue1:amd64 (0.4-2ubuntu1) ...\nSelecting previously unselected package libnfsidmap2:amd64.\nPreparing to unpack ...\/libnfsidmap2_0.25-5_amd64.deb ...\nUnpacking libnfsidmap2:amd64 (0.25-5) ...\nSelecting previously unselected package libtirpc1:amd64.\nPreparing to unpack ...\/libtirpc1_0.2.2-5ubuntu2_amd64.deb ...\nUnpacking libtirpc1:amd64 (0.2.2-5ubuntu2) ...\nSelecting previously unselected package keyutils.\nPreparing to unpack ...\/keyutils_1.5.6-1_amd64.deb ...\nUnpacking keyutils (1.5.6-1) ...\nSelecting previously unselected package rpcbind.\nPreparing to unpack ...\/rpcbind_0.2.1-2ubuntu2.2_amd64.deb ...\nUnpacking rpcbind (0.2.1-2ubuntu2.2) ...\nSelecting previously unselected package nfs-common.\nPreparing to unpack ...\/nfs-common_1%3a1.2.8-6ubuntu1.2_amd64.deb ...\nUnpacking nfs-common (1:1.2.8-6ubuntu1.2) ...\nProcessing triggers for man-db (2.6.7.1-1ubuntu1) ...\nProcessing triggers for ureadahead (0.100.0-16) ...\nSetting up libgssglue1:amd64 (0.4-2ubuntu1) ...\nSetting up libnfsidmap2:amd64 (0.25-5) ...\nSetting up libtirpc1:amd64 (0.2.2-5ubuntu2) ...\nSetting up keyutils (1.5.6-1) ...\nSetting up rpcbind (0.2.1-2ubuntu2.2) ...\n Removing any system startup links for \/etc\/init.d\/rpcbind ...\nrpcbind start\/running, process 2842\nProcessing triggers for ureadahead (0.100.0-16) ...\nSetting up nfs-common (1:1.2.8-6ubuntu1.2) ...\nAdding system user `statd' (UID 110) ...\nAdding new user `statd' (UID 110) with group `nogroup' ...\nNot creating home directory `\/var\/lib\/nfs'.\nstatd start\/running, process 3067\ngssd stop\/pre-start, process 3101\nidmapd start\/running, process 3151\nProcessing triggers for libc-bin (2.19-0ubuntu6.7) ...\nProcessing triggers for ureadahead (0.100.0-16) ...\n",
      "stdout_lines": [
        "Reading package lists...",
        "Building dependency tree...",
        "Reading state information...",
        "The following extra packages will be installed:",
        "  keyutils libgssglue1 libnfsidmap2 libtirpc1 rpcbind",
        "Suggested packages:",
        "  open-iscsi watchdog",
        "The following NEW packages will be installed:",
        "  keyutils libgssglue1 libnfsidmap2 libtirpc1 nfs-common rpcbind",
        "0 upgraded, 6 newly installed, 0 to remove and 29 not upgraded.",
        "Need to get 375 kB of archives.",
        "After this operation, 1524 kB of additional disk space will be used.",
        "Get:1 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libgssglue1 amd64 0.4-2ubuntu1 [19.7 kB]",
        "Get:2 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libnfsidmap2 amd64 0.25-5 [32.2 kB]",
        "Get:3 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main libtirpc1 amd64 0.2.2-5ubuntu2 [71.3 kB]",
        "Get:4 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty\/main keyutils amd64 1.5.6-1 [33.6 kB]",
        "Get:5 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty-updates\/main rpcbind amd64 0.2.1-2ubuntu2.2 [37.1 kB]",
        "Get:6 http:\/\/us-west-2.ec2.archive.ubuntu.com\/ubuntu\/ trusty-updates\/main nfs-common amd64 1:1.2.8-6ubuntu1.2 [181 kB]",
        "Fetched 375 kB in 0s (6884 kB\/s)",
        "Selecting previously unselected package libgssglue1:amd64.",
        "(Reading database ... 74563 files and directories currently installed.)",
        "Preparing to unpack ...\/libgssglue1_0.4-2ubuntu1_amd64.deb ...",
        "Unpacking libgssglue1:amd64 (0.4-2ubuntu1) ...",
        "Selecting previously unselected package libnfsidmap2:amd64.",
        "Preparing to unpack ...\/libnfsidmap2_0.25-5_amd64.deb ...",
        "Unpacking libnfsidmap2:amd64 (0.25-5) ...",
        "Selecting previously unselected package libtirpc1:amd64.",
        "Preparing to unpack ...\/libtirpc1_0.2.2-5ubuntu2_amd64.deb ...",
        "Unpacking libtirpc1:amd64 (0.2.2-5ubuntu2) ...",
        "Selecting previously unselected package keyutils.",
        "Preparing to unpack ...\/keyutils_1.5.6-1_amd64.deb ...",
        "Unpacking keyutils (1.5.6-1) ...",
        "Selecting previously unselected package rpcbind.",
        "Preparing to unpack ...\/rpcbind_0.2.1-2ubuntu2.2_amd64.deb ...",
        "Unpacking rpcbind (0.2.1-2ubuntu2.2) ...",
        "Selecting previously unselected package nfs-common.",
        "Preparing to unpack ...\/nfs-common_1%3a1.2.8-6ubuntu1.2_amd64.deb ...",
        "Unpacking nfs-common (1:1.2.8-6ubuntu1.2) ...",
        "Processing triggers for man-db (2.6.7.1-1ubuntu1) ...",
        "Processing triggers for ureadahead (0.100.0-16) ...",
        "Setting up libgssglue1:amd64 (0.4-2ubuntu1) ...",
        "Setting up libnfsidmap2:amd64 (0.25-5) ...",
        "Setting up libtirpc1:amd64 (0.2.2-5ubuntu2) ...",
        "Setting up keyutils (1.5.6-1) ...",
        "Setting up rpcbind (0.2.1-2ubuntu2.2) ...",
        " Removing any system startup links for \/etc\/init.d\/rpcbind ...",
        "rpcbind start\/running, process 2842",
        "Processing triggers for ureadahead (0.100.0-16) ...",
        "Setting up nfs-common (1:1.2.8-6ubuntu1.2) ...",
        "Adding system user `statd' (UID 110) ...",
        "Adding new user `statd' (UID 110) with group `nogroup' ...",
        "Not creating home directory `\/var\/lib\/nfs'.",
        "statd start\/running, process 3067",
        "gssd stop\/pre-start, process 3101",
        "idmapd start\/running, process 3151",
        "Processing triggers for libc-bin (2.19-0ubuntu6.7) ...",
        "Processing triggers for ureadahead (0.100.0-16) ..."
      ]
    },
    "exc_info": null,
    "exc_text": null,
    "filename": "cumulus_log.py",
    "funcName": "log",
    "levelname": "INFO",
    "levelno": 20,
    "lineno": 60,
    "module": "cumulus_log",
    "msecs": 283.99205207825,
    "msg": "TASK: exec : Install NFS client",
    "name": "cumulus_log",
    "pathname": "\/opt\/hpccloud\/cumulus\/cumulus\/ansible\/tasks\/playbooks\/callback_plugins\/cumulus_log.py",
    "process": 31766,
    "processName": "MainProcess",
    "relativeCreated": 372523.19598198,
    "status": "finished",
    "thread": 1.3965367359674e+14,
    "threadName": "MainThread",
    "type": "task"
  }
```

**Comment by TristanWright** [comment]
Identical messages have very close timestamps (µs apart). Can't find that we're appending twice anywhere could we be getting dup signals from celery? 

```json 
[
  {
    "args": [
      
    ],
    "created": 1467069776.4931,
    "data": {
      "_ansible_no_log": false,
      "changed": true,
      "module_name": "service",
      "name": "gridengine-exec",
      "state": "started"
    },
    "exc_info": null,
    "exc_text": null,
    "filename": "cumulus_log.py",
    "funcName": "log",
    "levelname": "INFO",
    "levelno": 20,
    "lineno": 60,
    "module": "cumulus_log",
    "msecs": 493.1321144104,
    "msg": "TASK: exec : Start execd",
    "name": "cumulus_log",
    "pathname": "\/opt\/hpccloud\/cumulus\/cumulus\/ansible\/tasks\/playbooks\/callback_plugins\/cumulus_log.py",
    "process": 31766,
    "processName": "MainProcess",
    "relativeCreated": 355732.33604431,
    "status": "finished",
    "thread": 1.3965367359674e+14,
    "threadName": "MainThread",
    "type": "task"
  },
  {
    "args": [
      
    ],
    "created": 1467069776.5161,
    "data": {
      "_ansible_no_log": false,
      "changed": true,
      "module_name": "service",
      "name": "gridengine-exec",
      "state": "started"
    },
    "exc_info": null,
    "exc_text": null,
    "filename": "cumulus_log.py",
    "funcName": "log",
    "levelname": "INFO",
    "levelno": 20,
    "lineno": 60,
    "module": "cumulus_log",
    "msecs": 516.0961151123,
    "msg": "TASK: exec : Start execd",
    "name": "cumulus_log",
    "pathname": "\/opt\/hpccloud\/cumulus\/cumulus\/ansible\/tasks\/playbooks\/callback_plugins\/cumulus_log.py",
    "process": 31766,
    "processName": "MainProcess",
    "relativeCreated": 355755.30004501,
    "status": "finished",
    "thread": 1.3965367359674e+14,
    "threadName": "MainThread",
    "type": "task"
  }
]
```

**Comment by cjh1** [comment]
I will take a look

File: girder/taskflow/server/tasks.py [hunk 2]
```diff
@@ -110,15 +107,10 @@ def log(self, task, params):
...
-            'The task to get log entries for.', paramType='path')
-        .param(
-            'offset',
-            'A offset in to the log.', required=False,
-            paramType='query')
+        Description('Get log entries for task')
+        .param('id', 'The task to get log entries for.', paramType='path')
+        .param('offset', 'A offset in to the log.', required=False,
+               paramType='query')
     )
```
Comment (cjh1) [inline girder/taskflow/server/tasks.py]: Any reason for this reformatting?

File: girder/taskflow/plugin.cmake [hunk 3]
```diff
@@ -1,2 +1,4 @@
-add_python_test(task PLUGIN task)
-add_python_style_test(pep8_style_task "${PROJECT_SOURCE_DIR}/plugins/task/server")
\ No newline at end of file
+add_python_test(task PLUGIN taskflow)
+add_python_test(taskflow PLUGIN taskflow)
+add_python_style_test(python_static_analysis_taskflow
+"${PROJECT_SOURCE_DIR}/plugins/taskflow/server")
```
Comment (cjh1) [inline girder/taskflow/plugin.cmake]: Don't think we need the line break here?

File: girder/taskflow/plugin_tests/taskflow_test.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright 2015 Kitware Inc.
+#
```
Comment (cjh1) [inline girder/taskflow/plugin_tests/taskflow_test.py]: 2016

File: girder/taskflow/server/tasks.py (hunk 2 above)
Comment (TristanWright) [inline girder/taskflow/server/tasks.py]: It definitely escapes me right now, I'm changing it back.

File: girder/taskflow/plugin.cmake (hunk 3 above)
Comment (TristanWright) [inline girder/taskflow/plugin.cmake]: reverted

File: girder/taskflow/plugin_tests/taskflow_test.py (hunk 4 above)
Comment (TristanWright) [inline girder/taskflow/plugin_tests/taskflow_test.py]: good catch, updated

**Comment by cjh1** [comment]
The duplicate message, are simply the same operation be performed on the different nodes in the cluster, so this is working as designed.

**Comment by cjh1** [comment]
LGTM

**Pull request merged**